### ISSUE TITLE
Discusion 4.0.x Way of improving generic database setup

### ISSUE THREAD

**Issue by dschissler** [phalcon/cphalcon#13647]
Here is an example of a typical generic database setup for a Phalcon 3 base project.  Notice that for Postgres the charset needs to be unset or apparently something bad will happen.  I'm thinking that it would be very nice if the builtin host based RDBSs didn't need this and would also default to utf8 by default.  I think that going forward about 99.5% of people want utf8 and fine if they don't then they can set it themselves.  It would be nice to do a bit of default handholding in Phalcon 4.  Does anyone else think that Phalcon should automatically set some PDO options and also to assume utf8 by default?

```php
    $class = $config->database->adapter;
    $params = [
        'host'     => $config->database->host,
        'username' => $config->database->username,
        'password' => $config->database->password,
        'dbname'   => $config->database->dbname,
        'charset'  => $config->database->charset,
    ];

    if ($class == Postgresql::class) {
        unset($params['charset']);
    }

    $connection = new $class($params);
```

**Comment by CameronHall** [comment]
I agree! Worth noting that for the MySQL adapter we should set the default charset to `utf8mb4.

**Comment by dschissler** [comment]
Oh right!  Good one.  Modern MySQL is pretty good but it sure has a funky history and PHP MySQL is even funkier.

**Comment by CameronHall** [comment]
It's both. The character set can be defined on the DB, table and the connection. When I create my PDO connection I do this;
```php
PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8mb4'
```

**Comment by CameronHall** [comment]
I'd never actually thought much about it until now. I found a [StackOverflow](https://stackoverflow.com/questions/24356090/difference-between-database-table-column-collation) answer explaining the order of precedence. 

> Order of precedence with topmost being least precedence:
> 
> - Server collation
> - Connection-specific collation
> - Database collation
> - Table collation
> - Column collation
> - Query collation (using CAST or CONVERT)

So I suppose setting the character set on the connection doesn't really do a whole lot when it can be overriden by everything else anyway (and rightfully so).

**Comment by dschissler** [comment]
@niden @sergeyklay I propose that this be added to the v4 project.  Since Phalcon 4 is not an LTS there can be some more wild experimentation here on saner PDO defaults.  This could also be kicked to Phalcon 5 since that will probably have tons of repository pattern work and so maybe it will better suited since that code is somewhat related to the DB work.  To reiterate what is said above, I'm referring to setting sane PDO defaults for modern MySQL and Postgres as well as possible better and more modern defaults as are set with `Mvc::setup`.  So the MySQL connection gets set to `utf8mb4`, PostgresSQL gets set to utf8-whatever, etc.  Eventually I'd like for MySQL values to be auto-casted in whatever appropriate layer, but for now perhaps we can just put in some quick and easy improved defaults.  In particular, now that `mysqli` is the only option my MySQL there is perhaps some low hanging fruit there.  I've written at least two hatred and curse word filled rants on the forum about this over the last 3 years and they have always been popular with thumbs up.  Sometimes its time to let loose the fury and instead of me being a utter maniac its appropriate due to how reviled the current situation is.  The king must die!

**Comment by niden** [comment]
@dschissler That is a nice enhancement. I think we can target this for `4.1` 

If anyone has time to research as to what the equivalents are for Postgresql and Sqlite (if any) that would be great

**Comment by niden** [comment]
I don't think I can squeeze this into 4.0 honestly. We will see. Let me provisionally add it in and see how it goes.

**Comment by dschissler** [comment]
Thanks @niden for merging Pdo radical options.

Now I'd like to propose something similar for setting post options.  Here is my updated comment from the last merged PR.

-----------------------------------

Next up is to add something like a `postOptions` setting.  I'm not 1000% sure but I found and read that Sqlite doesn't honour the option errors as exceptions when its set in the PDO constructor.  Its been many months since I've been able to verify it but it still feels trueish to me.

So this zephir line is not honored when using Sqlite.

```zephir
let options[\Pdo::ATTR_ERRMODE] = \Pdo::ERRMODE_EXCEPTION;
```

I had to fix it in my service definiion using the following:

```php
$db =  new Sqlite([
    'dbname'  => $filePath,
]);

$db->getInternalHandler()
    ->setAttribute(\PDO::ATTR_ERRMODE, \PDO::ERRMODE_EXCEPTION);
```

So I'm thinking about another PR `postOptions` (with defaults like dsn).  I feel that this is a low risk addition and if my reading and real world findings are true then Phalcon users would benefit tremendously from having errors as exceptions in **reality**.

I think that its too ambitious to attempt sql calls for things that just can't be setup otherwise like charset in postgres, etc.  Maybe Phalcon 5 or perhaps later.  It would be great if we could get a few of these in and to have a framework for adding another weird setting in the future if it comes up.

Basically to summarize, I would get the "internal pdo handler" and iterate on `postOptions` to call with `PDO::setAttribute`.
